### PR TITLE
Remove DP

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Removes DataParallel as it is [discouraged by PyTorch](https://github.com/pytorch/pytorch/issues/65936) and requires a lot of workarounds due to a different data-flow.

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

cc @justusschock @awaelchli @borda

### CODE REVIEW

**Review by awaelchli — APPROVED** [review]

File: docs/source-pytorch/guides/speed.rst [hunk 1]
```diff
@@ -49,19 +49,6 @@ GPU Training Speedup Tips
...
-3. Copy the outputs of each device back to the main device.
-
-.. image:: https://pl-public-data.s3.amazonaws.com/docs/static/images/distributed_training/dp.gif
-    :alt: Animation showing DP execution.
-    :width: 500
-    :align: center
-
 |
 
 Whereas :class:`~pytorch_lightning.strategies.ddp.DDPStrategy` only performs two transfer operations, making DDP much faster than DP:
```
Comment (carmocca) [inline docs/source-pytorch/guides/speed.rst:54]: This needs to be updated

File: src/lightning/pytorch/CHANGELOG.md [hunk 2]
```diff
@@ -284,6 +284,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 - Removed support for passing a scheduling dictionary to `Trainer(accumulate_grad_batches=...)` ([#16729](https://github.com/Lightning-AI/lightning/pull/16729))
 
+
+- Removed support for `DataParallel` (`strategy='dp'`) ([#16748](https://github.com/Lightning-AI/lightning/pull/16748))
```
Comment (carmocca) [inline src/lightning/pytorch/CHANGELOG.md:288]: Can you mention the removed classes in `src/lightning/pytorch/overrides/data_parallel`

File: tests/tests_pytorch/utilities/test_compile.py [hunk 3]
```diff
@@ -57,13 +56,6 @@ def test_trainer_compiled_model(tmp_path, monkeypatch):
     trainer.fit(model)
     assert trainer.model._compiler_ctx is None
 
-    # some strategies do not support it
-    compiled_model = torch.compile(model)
-    mock_cuda_count(monkeypatch, 1)
-    trainer = Trainer(strategy="dp", accelerator="cuda", **trainer_kwargs)
```
Comment (carmocca) [inline tests/tests_pytorch/utilities/test_compile.py:63]: This should be kept but using a different strategy that is not supported with `torch.compile`

**Review by carmocca — APPROVED** [review]
Pushed some missed removals

File: docs/source-pytorch/accelerators/gpu_intermediate.rst [hunk 4]
```diff
@@ -234,107 +210,26 @@ Comparison of DDP variants and tradeoffs
...
+| Y     |         |     |        | `Trainer(accelerator="gpu", devices=1)`                               |
++-------+---------+-----+--------+-----------------------------------------------------------------------+
+| Y     |         |     | Y      | `Trainer(accelerator="gpu", devices=1, precision=16)`                 |
++-------+---------+-----+--------+-----------------------------------------------------------------------+
+|       | Y       | Y   |        | `Trainer(accelerator="gpu", devices=k, strategy='ddp')`               |
++-------+---------+-----+--------+-----------------------------------------------------------------------+
+|       | Y       | Y   | Y      | `Trainer(accelerator="gpu", devices=k, strategy='ddp', precision=16)` |
++-------+---------+-----+--------+-----------------------------------------------------------------------+
+|       | Y       |     |        | `Trainer(accelerator="gpu", devices=k, strategy='dp')`                |
++-------+---------+-----+--------+-----------------------------------------------------------------------+
```
Comment (awaelchli) [inline docs/source-pytorch/accelerators/gpu_intermediate.rst:230]:
```suggestion
+-------+---------+-----+--------+-----------------------------------------------------------------------+
```

**Review by awaelchli — APPROVED** [review]
